src/cli.rs

```diff
@@ -1,88 +1,156 @@
+    toolchains: Vec<String>,
+
+    /// Turn verbose information on
+    #[arg(short, long)]
+    pub verbose: bool,
+}
+
+impl Args {
+    pub fn criterion(&self) -> Criterion {
+        match &self {
+            _ if self.stamp => Criterion::Stamp,
+            _ if self.file => Criterion::File,
+            _ if self.installed => Criterion::Installed,
+            _ if !self.toolchains.is_empty() => Criterion::Toolchains(self.toolchains.clone()),
+            Self {
+                time: Some(time), ..
+            } => Criterion::Time(*time),
+            Self {
+                maxsize: Some(size),
+                ..
+            } => Criterion::MaxSize(*size),
+            _ => unreachable!("guaranteed by clap ArgGroup"),
+        }
+    }
+}
+
+pub enum Criterion {
+    Stamp,
+    File,
+    Time(u64),
+    Installed,
+    Toolchains(Vec<String>),
+    MaxSize(u64),
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    /// Test helper for splitting arguments and providing them to clap
+    fn parse(command: &str) -> Result<Args, clap::Error> {
+        let command_args = command.split_whitespace();
+        dbg!(SweepArgs::try_parse_from(command_args).map(SweepArgs::into_args))
+    }
+
+    #[test]
+    fn test_argparsing() {
+        // Argument is required
+        assert!(parse("cargo sweep").is_err());
+        assert!(parse("cargo sweep --installed").is_ok());
+        assert!(parse("cargo sweep --file").is_ok());
+        assert!(parse("cargo sweep --stamp").is_ok());
+        assert!(parse("cargo sweep --time 30").is_ok());
+        assert!(parse("cargo sweep --toolchains SAMPLE_TEXT").is_ok());
+        assert!(parse("cargo sweep --maxsize 100").is_ok());
+
+        assert!(parse("cargo-sweep sweep").is_err());
+        assert!(parse("cargo-sweep sweep --installed").is_ok());
+        assert!(parse("cargo-sweep sweep --file").is_ok());
+        assert!(parse("cargo-sweep sweep --stamp").is_ok());
+        assert!(parse("cargo-sweep sweep --time 30").is_ok());
+        assert!(parse("cargo-sweep sweep --toolchains SAMPLE_TEXT").is_ok());
+        assert!(parse("cargo-sweep sweep --maxsize 100").is_ok());
+
+        // Argument conflicts
+        assert!(parse("cargo sweep --installed --maxsize 100").is_err());
+        assert!(parse("cargo sweep --file --installed").is_err());
+        assert!(parse("cargo sweep --stamp --file").is_err());
+        assert!(parse("cargo sweep --time 30 --stamp").is_err());
+        assert!(parse("cargo sweep --toolchains SAMPLE_TEXT --time 30").is_err());
+        assert!(parse("cargo sweep --maxsize 100 --toolchains SAMPLE_TEXT").is_err());
+
+        // Test if comma separated list is parsed correctly
+        let args = Args {
+            toolchains: ["1", "2", "3"].map(ToString::to_string).to_vec(),
+            ..Args::default()
+        };
+        assert_eq!(args, parse("cargo sweep --toolchains 1,2,3").unwrap());
+        assert!(parse("cargo sweep --toolchains 1 2 3").is_err());
```

Comment:
This doesn't look right - it's valid to have toolchains with spaces.
```
(-zsh@jyns-MacBook-Pro) ~/src/cargo-sweep [16:33:23]
; rustup toolchain link 'toolchain with spaces' ~/src/rust/build/x86_64-apple-darwin/stage0
(-zsh@jyns-MacBook-Pro) ~/src/cargo-sweep [16:33:38]
; rustc '+toolchain with spaces' --version
rustc 1.67.0-beta.3 (b364405cc 2022-12-27)
```
(I guess it's valid to have commas too, but that's a pre-existing issue, seems ok to not handle for now.)

Reply:
`'+toolchain with spaces'` is just one arg, while in this test, each arg is treated as a different one (because of how the helper `parse` is implemented).

In other words, this test is testing for something like:
```
rustc +toolchain with spaces --version
```
Instead of:
```
rustc '+toolchain with spaces' --version
```
Do you think I should change that?

Reply:
ah, got it. no need to change it - turns out rustup doesn't handle spaces properly anyway: https://github.com/rust-lang/rustup/issues/3168